trpcultivate_phenotypes/tests/src/Kernel/Validators/ValidatorValidHeadersTest.php

```diff
@@ -0,0 +1,377 @@
+            'Header 5',
+            'Header 4',
+            'Header 3',
+            'Header 2',
+            'Header 1',
+            'Header 0',
+          ]
+        ],
+        [
+          'number_of_columns' => 6,
+          'strict' => FALSE
+        ]
+      ],
+
+      // #6: A valid header but the expected column strict comparison flag is set to exact match (TRUE).
+      [
+        'not have expected number',
+        [
+          'Header 0',
+          'Header 1',
+          'Header 2',
+          'Header 3',
+          'Header 4',
+          'Header 5'
+        ],
+        [
+          'case' => 'Headers provided does not have the expected number of headers',
+          'valid' => FALSE,
+          'failedItems' => [
+            'Header 0',
+            'Header 1',
+            'Header 2',
+            'Header 3',
+            'Header 4',
+            'Header 5',
+          ]
+        ],
+        [
+          'number_of_columns' => 50,
+          'strict' => TRUE
+        ]
+      ],
+
+      // #7: Missing header with string index values.
+      [
+        'string index invalid',
+        [
+          'zero' => 'Header 0',
+          'one' => 'Header 1',
+          'three' => 'Header 3',
+          'four' => 'Header 4',
+          'five' => 'Header 5'
+        ],
+        [
+          'case' => 'Headers do not match expected headers',
+          'valid' => FALSE,
+          'failedItems' => [
+            'zero' => 'Header 0',
+            'one' => 'Header 1',
+            'three' => 'Header 3',
+            'four' => 'Header 4',
+            'five' => 'Header 5'
+          ]
+        ],
+        [
+          'number_of_columns' => 6,
+          'strict' => FALSE
+        ]
+      ],
+
+      // #8: A valid header with string index values.
+      [
+        'string index valid',
+        [
+          'zero' => 'Header 0',
+          'one' => 'Header 1',
+          'two' => 'Header 2',
+          'three' => 'Header 3',
+          'four' => 'Header 4',
+          'five' => 'Header 5'
+        ],
+        [
+          'case' => 'Headers exist and match expected headers',
+          'valid' => TRUE,
+          'failedItems' => []
+        ],
+        [
+          'number_of_columns' => 6,
+          'strict' => FALSE
+        ]
+      ],
+
+      // #9: A valid header (list and order).
+      [
+        'valid headers',
+        [
+          'Header 0',
+          'Header 1',
+          'Header 2',
+          'Header 3',
+          'Header 4',
+          'Header 5'
+        ],
+        [
+          'case' => 'Headers exist and match expected headers',
+          'valid' => TRUE,
+          'failedItems' => []
+        ],
+        [
+          'number_of_columns' => 6,
+          'strict' => FALSE
+        ]
+      ],
+    ];
+  }
+
+  /**
+   * Test headers validator.
+   *
+   * @dataProvider provideHeadersToHeadersValidator
+   */
+  public function testHeaders($scenario, $headers_input, $expected, $expected_columns) {
```

Comment:
It would be good to check that the failedItems in expected match the headers input just as a check that you setup your scenarios properly.

And maybe also check the order matches ;-p